secure/json-rpc/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+use ureq::Response;
+
+mod error;
+#[cfg(test)]
+mod tests;
+
+/// Various constants for the JSON RPC client implementation
+const JSON_RPC_VERSION: &str = "2.0";
+const REQUEST_TIMEOUT: u64 = 10_000;
+
+/// This defines a simple and generic interface by which to interact with the JSON RPC service
+/// offered by each Libra Full Node.
+pub trait JsonRpcClient {
+    /// Submits a signed transaction to the Libra blockchain via the JSON RPC API.
+    fn submit_signed_transaction(&self, signed_transaction: SignedTransaction)
+        -> Result<(), Error>;
+
+    /// Returns the associated AccountState for a specific account at a given version height
+    /// using the JSON RCP API.
+    fn get_account_state(
+        &self,
+        account: AccountAddress,
+        version: u64,
+    ) -> Result<AccountState, Error>;
+}
+
+/// Provides a lightweight JsonRpcClient implementation.
+struct LightweightJsonClient {
+    host: String,
+}
+
+impl LightweightJsonClient {
+    // TODO(joshlind): remove this flag once the client is integrated into key manager. Otherwise,
+    // rust complains new() is never used.
+    #[cfg(test)]
+    pub fn new(host: &str, port: u16) -> Self {
+        let host = format!("http://{}:{}", host, port);
+        Self { host }
+    }
+
+    // Executes the specified request method using the given parameters by contacting the JSON RPC
+    // server.
+    fn execute_request(&self, method: String, params: Vec<Value>) -> Response {
+        let request_id = 0;
+        ureq::post(&self.host)
+            .timeout_connect(REQUEST_TIMEOUT)
+            .send_json(
+                json!({"jsonrpc": JSON_RPC_VERSION, "method": method, "params": params, "id": request_id}),
+            )
+    }
+
+    // Sends a submit() request to the JSON RPC server using the given transaction.
+    fn submit(&self, signed_transaction: SignedTransaction) -> Result<Response, Error> {
+        let method = "submit".to_string();
+        let params = vec![Value::String(hex::encode(lcs::to_bytes(
+            &signed_transaction,
+        )?))];
+        Ok(self.execute_request(method, params))
+    }
+
+    // Sends a get_account_state_with_proof() request to the JSON RPC server for the specified
+    // account, version height and ledger_version height (for the proof).
+    fn get_account_state_with_proof(
+        &self,
+        account: AccountAddress,
+        version: u64,
+        ledger_version: u64,
+    ) -> Result<Response, Error> {
+        let method = "get_account_state_with_proof".to_string();
+        let params = vec![
+            Value::String(account.to_string()),
+            Value::Number(version.into()),
+            Value::Number(ledger_version.into()),
+        ];
+        Ok(self.execute_request(method, params))
+    }
+
+    /// Parses the response returned by the submit() JSON RPC request, and ensures the request
+    /// was successful.
+    fn process_submit_response(&self, response: Response) -> Result<(), Error> {
```

Comment:
is there a reason to not have this inline where it is being used only once? It makes it easier to review the code.

Reply:
As discussed offline, this was more a readability thing (as some of the functions are pretty long/nested otherwise), but I guess that seems to be a result of the way Rust works sometimes.. I've inlined the functions.. let me know what you think. Personally I think the 'get_account_state' function is a little long, but it's not super bad!